zebra-chain/src/transaction.rs

```diff
@@ -76,44 +76,133 @@ pub struct TransparentOutput {
+/// designed to ensure that transactions can created, propagated on the network,
+/// validated, and finally added to the global ledger of transactions (the
+/// blockchain).
+///
+/// Zcash has a number of different transaction formats. They are represented
+/// internally by different enum variants. Because we checkpoint on Sapling
+/// activation, we do not parse any pre-Sapling transaction types.
+#[derive(Clone, Debug, PartialEq, Eq)]
+pub enum Transaction {
+    /// A fully transparent transaction (`version = 1`).
+    Transparent {
+        /// The transparent inputs to the transaction.
+        inputs: Vec<TransparentInput>,
+        /// The transparent outputs from the transaction.
+        outputs: Vec<TransparentOutput>,
+        /// The earliest time or block height that this transaction can be added to the
+        /// chain.
+        lock_time: LockTime,
+    },
+    /// A Sprout transaction (`version = 2`).
+    Sprout {
+        /// The transparent inputs to the transaction.
+        inputs: Vec<TransparentInput>,
+        /// The transparent outputs from the transaction.
+        outputs: Vec<TransparentOutput>,
+        /// The earliest time or block height that this transaction can be added to the
+        /// chain.
+        lock_time: LockTime,
+        /// The JoinSplit data for this transaction, if any.
+        js_data: Option<LegacyJoinSplitData>,
+    },
+    /// An Overwinter transaction (`version = 3`).
+    Overwinter {
+        /// The transparent inputs to the transaction.
+        inputs: Vec<TransparentInput>,
+        /// The transparent outputs from the transaction.
+        outputs: Vec<TransparentOutput>,
+        /// The earliest time or block height that this transaction can be added to the
+        /// chain.
+        lock_time: LockTime,
+        /// The latest block height that this transaction can be added to the chain.
+        expiry_height: BlockHeight,
+        /// The JoinSplit data for this transaction, if any.
+        js_data: Option<LegacyJoinSplitData>,
+    },
+    /// A Sapling transaction (`version = 4`).
+    Sapling {
+        /// The transparent inputs to the transaction.
+        inputs: Vec<TransparentInput>,
+        /// The transparent outputs from the transaction.
+        outputs: Vec<TransparentOutput>,
+        /// The earliest time or block height that this transaction can be added to the
+        /// chain.
+        lock_time: LockTime,
+        /// The latest block height that this transaction can be added to the chain.
+        expiry_height: BlockHeight,
+        /// The net value of Sapling spend transfers minus output transfers.
+        // XXX should this be an Amount type?
+        value_balance: i64,
+        /// The shielded data for this transaction, if any.
+        shielded_data: Option<ShieldedData>,
+        /// The JoinSplit data for this transaction, if any.
+        js_data: Option<SaplingJoinSplitData>,
```

Comment:
This approach to ensuring structural integrity seems legit. There will be lots of cases where you don't actually care what version a tx is as long as it has a particular field, though, so there should probably also be methods to return each field regardless of version.

Reply:
Yep, this is a great idea.